File: Advance-Data-Structure/PYTHON/stack.py
class Stack:
    def __init__(self):
        self.items = []
        
    def is_empty(self):
        # return len(self.items) == 0
        return not self.items
    
    def push(self, item):
        self.items.append(item)
        
    def pop(self):
        return self.items.pop()
    
    def top(self):
        return self.items[-1]
    
    def size(self):
        return len(self.items)
    
    def __str__(self):
        return str(self.items)
    
if __name__ == '__main__':
    s = Stack()
    print(s)
    print(s.is_empty())
    s.push(1)
    s.push(2)
    print(s)
    s.push(3)
    print(s)
    print(s.top())
    print(s.pop())
    print(s)
    
"""
[]
True
[1, 2]
[1, 2, 3]
3
3
[1, 2]
"""

File: Advance-Data-Structure/Stack.cpp
#include <bits/stdc++.h>
using namespace std;
#define SIZE 10

typedef struct {
    int top;
    int arr[SIZE];
} Stack;

void push(Stack *st, int item) {
    if (st->top == SIZE) {
        printf("Error: Stack is full\n");
        return;
    }
    st->arr[st->top] = item;
    st->top = st->top + 1;
}

int pop(Stack *st) {
    if (st->top == 0) {
        printf("Error: Stack is empty\n");
        exit(EXIT_FAILURE);
    }
    st->top = st->top - 1;
    return st->arr[st->top];
}

int main() {
    Stack mystack;
    mystack.top = 0;
    push(&mystack, 1);
    push(&mystack, 2);
    push(&mystack, 3);
    printf("%d\n", pop(&mystack));
    printf("%d\n", pop(&mystack));
    printf("%d\n", pop(&mystack));
    printf("%d\n", pop(&mystack));
    return 0;
}
